File: analyzers/js-ts/src/ast-walker.ts
import ts from "typescript";
import type {
  FunctionIR,
  CallIR,
  ImportIR,
  ExportIR,
  ClassIR,
  ParameterIR,
  SourceLocation,
  ImportedSymbolIR,
} from "./types.js";
import type { FrameworkPlugin } from "./framework-plugins/index.js";

export interface FileWalkResult {
  functions: FunctionIR[];
  calls: CallIR[];
  imports: ImportIR[];
  exports: ExportIR[];
  classes: ClassIR[];
}

function getLocation(node: ts.Node, sourceFile: ts.SourceFile): SourceLocation {
  const start = sourceFile.getLineAndCharacterOfPosition(node.getStart(sourceFile));
  const end = sourceFile.getLineAndCharacterOfPosition(node.getEnd());
  return {
    startLine: start.line + 1,
    endLine: end.line + 1,
    startColumn: start.character + 1,
    endColumn: end.character + 1,
  };
}

function getParameterIR(param: ts.ParameterDeclaration, sourceFile: ts.SourceFile): ParameterIR {
  const name = param.name.getText(sourceFile);
  const typeAnnotation = param.type ? param.type.getText(sourceFile) : null;
  const hasDefault = param.initializer !== undefined;
  const isRest = param.dotDotDotToken !== undefined;
  return { name, typeAnnotation, hasDefault, isRest };
}

function getReturnType(
  node: ts.FunctionDeclaration | ts.MethodDeclaration | ts.ArrowFunction | ts.FunctionExpression,
  sourceFile: ts.SourceFile
): string | null {
  if (node.type) {
    return node.type.getText(sourceFile);
  }
  return null;
}

function hasModifier(node: ts.Node, kind: ts.SyntaxKind): boolean {
  const modifiers = ts.canHaveModifiers(node) ? ts.getModifiers(node) : undefined;
  return modifiers?.some((m) => m.kind === kind) ?? false;
}

function isNodeExported(node: ts.Node): boolean {
  return hasModifier(node, ts.SyntaxKind.ExportKeyword);
}

function buildSignature(
  name: string,
  params: ParameterIR[],
  returnType: string | null
): string {
  const paramStr = params
    .map((p) => {
      let s = p.isRest ? `...${p.name}` : p.name;
      if (p.typeAnnotation) s += `: ${p.typeAnnotation}`;
      if (p.hasDefault) s += " = ...";
      return s;
    })
    .join(", ");
  const ret = returnType ? `: ${returnType}` : "";
  return `${name}(${paramStr})${ret}`;
}

function getCalleeText(expr: ts.Expression, sourceFile: ts.SourceFile): string {
  return expr.getText(sourceFile);
}

function getReceiverAndMethod(
  expr: ts.LeftHandSideExpression,
  sourceFile: ts.SourceFile
): { receiver: string | null; method: string | null } {
  if (ts.isPropertyAccessExpression(expr)) {
    return {
      receiver: expr.expression.getText(sourceFile),
      method: expr.name.getText(sourceFile),
    };
  }
  return { receiver: null, method: null };
}

function extractStringArgs(args: ts.NodeArray<ts.Expression>, sourceFile: ts.SourceFile): string[] {
  const result: string[] = [];
  for (const arg of args) {
    if (ts.isStringLiteral(arg) || ts.isNoSubstitutionTemplateLiteral(arg)) {
      result.push(arg.text);
    }
  }
  return result;
}

function extractArgumentRefs(args: ts.NodeArray<ts.Expression>, sourceFile: ts.SourceFile): string[] {
  const result: string[] = [];
  for (const arg of args) {
    if (ts.isIdentifier(arg)) {
      result.push(arg.text);
    }
  }
  return result;
}

export function walkFile(
  sourceFile: ts.SourceFile,
  program: ts.Program,
  plugins: FrameworkPlugin[]
): FileWalkResult {
  const functions: FunctionIR[] = [];
  const calls: CallIR[] = [];
  const imports: ImportIR[] = [];
  const exports: ExportIR[] = [];
  const classes: ClassIR[] = [];

  const enclosingFunctionStack: string[] = [];

  function currentEnclosingFunction(): string | null {
    return enclosingFunctionStack.length > 0
      ? enclosingFunctionStack[enclosingFunctionStack.length - 1]
      : null;
  }

  function processFunctionLike(
    node: ts.FunctionDeclaration | ts.MethodDeclaration | ts.ArrowFunction | ts.FunctionExpression,
    name: string,
    qualifiedName: string | null,
    exported: boolean,
    isStatic: boolean,
    accessibility: "public" | "protected" | "private" | null
  ): FunctionIR {
    const params = node.parameters.map((p) => getParameterIR(p, sourceFile));
    const returnType = getReturnType(node, sourceFile);
    const isAsync = hasModifier(node, ts.SyntaxKind.AsyncKeyword);
    const signature = buildSignature(name, params, returnType);
    const location = getLocation(node, sourceFile);

    const funcIR: FunctionIR = {
      kind: "function",
      name,
      qualifiedName,
      signature,
      parameters: params,
      returnType,
      isExported: exported,
      isAsync,
      isStatic: isStatic || undefined,
      accessibility: accessibility || undefined,
      location,
    };

    return funcIR;
  }

  function visitCallExpression(node: ts.CallExpression): void {
    const callee = getCalleeText(node.expression, sourceFile);
    const { receiver, method } = getReceiverAndMethod(node.expression, sourceFile);
    const stringArgs = extractStringArgs(node.arguments, sourceFile);
    const argumentRefs = extractArgumentRefs(node.arguments, sourceFile);

    const callIR: CallIR = {
      kind: "call",
      callee,
      receiver,
      method,
      argumentCount: node.arguments.length,
      argumentRefs: argumentRefs.length > 0 ? argumentRefs : undefined,
      stringArgs: stringArgs.length > 0 ? stringArgs : undefined,
      enclosingFunction: currentEnclosingFunction(),
      location: getLocation(node, sourceFile),
    };

    // Run framework plugins on calls
    for (const plugin of plugins) {
      if (plugin.analyzeCall) {
        const enrichment = plugin.analyzeCall(callIR, node, sourceFile);
        if (enrichment) {
          if (!callIR.enrichments) callIR.enrichments = [];
          callIR.enrichments.push(enrichment);
        }
      }
    }

    calls.push(callIR);
  }

  function visitImportDeclaration(node: ts.ImportDeclaration): void {
    if (!ts.isStringLiteral(node.moduleSpecifier)) return;

    const modulePath = node.moduleSpecifier.text;
    const isExternal = !modulePath.startsWith(".") && !modulePath.startsWith("/");

    let defaultImport: string | null = null;
    let namespaceImport: string | null = null;
    const symbols: ImportedSymbolIR[] = [];

    const importClause = node.importClause;
    if (importClause) {
      if (importClause.name) {
        defaultImport = importClause.name.text;
      }
      const bindings = importClause.namedBindings;
      if (bindings) {
        if (ts.isNamespaceImport(bindings)) {
          namespaceImport = bindings.name.text;
        } else if (ts.isNamedImports(bindings)) {
          for (const el of bindings.elements) {
            symbols.push({
              name: (el.propertyName || el.name).text,
              alias: el.propertyName ? el.name.text : null,
            });
          }
        }
      }
    }

    // Try to resolve the module path
    let resolvedPath: string | null = null;
    if (!isExternal) {
      const resolved = ts.resolveModuleName(
        modulePath,
        sourceFile.fileName,
        program.getCompilerOptions(),
        ts.sys
      );
      if (resolved.resolvedModule) {
        resolvedPath = resolved.resolvedModule.resolvedFileName;
      }
    }

    imports.push({
      kind: "import",
      modulePath,
      resolvedPath,
      isExternal,
      symbols,
      defaultImport,
      namespaceImport,
      location: getLocation(node, sourceFile),
    });
  }

  function visitExportDeclaration(node: ts.ExportDeclaration): void {
    const fromModule = node.moduleSpecifier && ts.isStringLiteral(node.moduleSpecifier)
      ? node.moduleSpecifier.text
      : null;

    if (node.exportClause && ts.isNamedExports(node.exportClause)) {
      for (const el of node.exportClause.elements) {
        const exportedName = el.name.text;
        const localName = el.propertyName ? el.propertyName.text : null;
        exports.push({
          kind: "export",
          name: exportedName,
          localName,
          isDefault: false,
          fromModule,
          location: getLocation(node, sourceFile),
        });
      }
    }
  }

  function visitExportAssignment(node: ts.ExportAssignment): void {
    const name = node.expression.getText(sourceFile);
    exports.push({
      kind: "export",
      name,
      localName: null,
      isDefault: !node.isExportEquals,
      fromModule: null,
      location: getLocation(node, sourceFile),
    });
  }

  function visitRequire(node: ts.VariableStatement): void {
    for (const decl of node.declarationList.declarations) {
      if (
        decl.initializer &&
        ts.isCallExpression(decl.initializer) &&
        ts.isIdentifier(decl.initializer.expression) &&
        decl.initializer.expression.text === "require" &&
        decl.initializer.arguments.length === 1 &&
        ts.isStringLiteral(decl.initializer.arguments[0])
      ) {
        const modulePath = (decl.initializer.arguments[0] as ts.StringLiteral).text;
        const isExternal = !modulePath.startsWith(".") && !modulePath.startsWith("/");
        const varName = decl.name.getText(sourceFile);

        imports.push({
          kind: "import",
          modulePath,
          resolvedPath: null,
          isExternal,
          symbols: [],
          defaultImport: null,
          namespaceImport: varName,
          location: getLocation(node, sourceFile),
        });
      }
    }
  }

  function visit(node: ts.Node): void {
    // Import declarations
    if (ts.isImportDeclaration(node)) {
      visitImportDeclaration(node);
      return;
    }

    // Export declarations (named exports)
    if (ts.isExportDeclaration(node)) {
      visitExportDeclaration(node);
      // Don't return - still visit children
    }

    // Export assignment (export default X)
    if (ts.isExportAssignment(node)) {
      visitExportAssignment(node);
      return;
    }

    // Require statements
    if (ts.isVariableStatement(node)) {
      visitRequire(node);
    }

    // Function declarations
    if (ts.isFunctionDeclaration(node) && node.name) {
      const name = node.name.text;
      const exported = isNodeExported(node);
      const funcIR = processFunctionLike(node, name, null, exported, false, null);

      if (exported) {
        exports.push({
          kind: "export",
          name,
          localName: null,
          isDefault: hasModifier(node, ts.SyntaxKind.DefaultKeyword),
          fromModule: null,
          location: getLocation(node, sourceFile),
        });
      }

      functions.push(funcIR);

      // Walk body
      enclosingFunctionStack.push(name);
      ts.forEachChild(node, visit);
      enclosingFunctionStack.pop();

      // Run plugins on the function after walking body (so calls are collected)
      for (const plugin of plugins) {
        if (plugin.analyzeFunction) {
          const result = plugin.analyzeFunction(funcIR, node, sourceFile, calls);
          if (result?.endpointInfo) funcIR.endpointInfo = result.endpointInfo;
          if (result?.enrichment) {
            if (!funcIR.enrichments) funcIR.enrichments = [];
            funcIR.enrichments.push(result.enrichment);
          }
        }
      }
      return;
    }

    // Variable declarations with arrow functions / function expressions
    if (ts.isVariableStatement(node)) {
      const exported = isNodeExported(node);
      for (const decl of node.declarationList.declarations) {
        if (
          decl.initializer &&
          ts.isIdentifier(decl.name) &&
          (ts.isArrowFunction(decl.initializer) || ts.isFunctionExpression(decl.initializer))
        ) {
          const name = decl.name.text;
          const funcIR = processFunctionLike(decl.initializer, name, null, exported, false, null);

          if (exported) {
            exports.push({
              kind: "export",
              name,
              localName: null,
              isDefault: false,
              fromModule: null,
              location: getLocation(node, sourceFile),
            });
          }

          functions.push(funcIR);

          enclosingFunctionStack.push(name);
          ts.forEachChild(decl.initializer, visit);
          enclosingFunctionStack.pop();

          // Run plugins
          for (const plugin of plugins) {
            if (plugin.analyzeFunction) {
              const result = plugin.analyzeFunction(funcIR, decl.initializer, sourceFile, calls);
              if (result?.endpointInfo) funcIR.endpointInfo = result.endpointInfo;
              if (result?.enrichment) {
                if (!funcIR.enrichments) funcIR.enrichments = [];
                funcIR.enrichments.push(result.enrichment);
              }
            }
          }
          continue;
        }
      }
      // Still visit children for other variable declarations
      ts.forEachChild(node, visit);
      return;
    }

    // Class declarations
    if (ts.isClassDeclaration(node)) {
      const className = node.name?.text ?? "<anonymous>";
      const exported = isNodeExported(node);
      const isAbstract = hasModifier(node, ts.SyntaxKind.AbstractKeyword);

      let superClass: string | null = null;
      const implementsList: string[] = [];

      if (node.heritageClauses) {
        for (const clause of node.heritageClauses) {
          if (clause.token === ts.SyntaxKind.ExtendsKeyword) {
            superClass = clause.types[0]?.expression.getText(sourceFile) ?? null;
          } else if (clause.token === ts.SyntaxKind.ImplementsKeyword) {
            for (const t of clause.types) {
              implementsList.push(t.expression.getText(sourceFile));
            }
          }
        }
      }

      const methodNames: string[] = [];

      // Visit class members
      for (const member of node.members) {
        if (ts.isMethodDeclaration(member) && member.name) {
          const methodName = member.name.getText(sourceFile);
          const qualifiedName = `${className}.${methodName}`;
          const isStatic = hasModifier(member, ts.SyntaxKind.StaticKeyword);

          let accessibility: "public" | "protected" | "private" | null = null;
          if (hasModifier(member, ts.SyntaxKind.PublicKeyword)) accessibility = "public";
          else if (hasModifier(member, ts.SyntaxKind.ProtectedKeyword)) accessibility = "protected";
          else if (hasModifier(member, ts.SyntaxKind.PrivateKeyword)) accessibility = "private";

          const funcIR = processFunctionLike(
            member,
            methodName,
            qualifiedName,
            exported,
            isStatic,
            accessibility
          );
          functions.push(funcIR);
          methodNames.push(methodName);

          enclosingFunctionStack.push(qualifiedName);
          ts.forEachChild(member, visit);
          enclosingFunctionStack.pop();
        } else if (ts.isConstructorDeclaration(member)) {
          const params = member.parameters.map((p) => getParameterIR(p, sourceFile));
          const signature = buildSignature("constructor", params, null);
          const funcIR: FunctionIR = {
            kind: "function",
            name: "constructor",
            qualifiedName: `${className}.constructor`,
            signature,
            parameters: params,
            returnType: null,
            isExported: exported,
            isAsync: false,
            location: getLocation(member, sourceFile),
          };
          functions.push(funcIR);
          methodNames.push("constructor");

          enclosingFunctionStack.push(`${className}.constructor`);
          ts.forEachChild(member, visit);
          enclosingFunctionStack.pop();
        }
      }

      if (exported) {
        exports.push({
          kind: "export",
          name: className,
          localName: null,
          isDefault: hasModifier(node, ts.SyntaxKind.DefaultKeyword),
          fromModule: null,
          location: getLocation(node, sourceFile),
        });
      }

      classes.push({
        kind: "class",
        name: className,
        superClass,
        implements: implementsList,
        isExported: exported,
        isAbstract,
        methods: methodNames,
        location: getLocation(node, sourceFile),
      });

      return;
    }

    // Call expressions
    if (ts.isCallExpression(node)) {
      visitCallExpression(node);
      // Continue visiting children (args may contain more calls)
      ts.forEachChild(node, visit);
      return;
    }

    ts.forEachChild(node, visit);
  }

  ts.forEachChild(sourceFile, visit);

  return { functions, calls, imports, exports, classes };
}

File: analyzers/go/analyzer/ast_walker.go
package analyzer

import (
	"fmt"
	"go/ast"
	"go/token"
	"strings"
	"unicode"

	"github.com/confianalyzer/analyzer-go/analyzer/plugins"
	"github.com/confianalyzer/analyzer-go/irtypes"
)

// WalkResult holds all IR data extracted from a single Go file.
type WalkResult struct {
	Functions []irtypes.FunctionIR
	Calls     []irtypes.CallIR
	Imports   []irtypes.ImportIR
	Exports   []irtypes.ExportIR
	Classes   []irtypes.ClassIR
}

// WalkFile extracts IR data from a parsed Go AST file.
func WalkFile(fset *token.FileSet, file *ast.File, modulePath string, registry *plugins.Registry) *WalkResult {
	w := &walker{
		fset:       fset,
		file:       file,
		modulePath: modulePath,
		registry:   registry,
	}
	w.walk()
	return &w.result
}

type walker struct {
	fset              *token.FileSet
	file              *ast.File
	modulePath        string
	registry          *plugins.Registry
	result            WalkResult
	enclosingFunction string
	structMethods     map[string][]string
}

func (w *walker) walk() {
	w.structMethods = make(map[string][]string)

	// First pass: collect all methods (FuncDecl with receivers) for struct method lists
	for _, decl := range w.file.Decls {
		if fd, ok := decl.(*ast.FuncDecl); ok && fd.Recv != nil {
			recvType := receiverTypeName(fd.Recv)
			if recvType != "" {
				w.structMethods[recvType] = append(w.structMethods[recvType], fd.Name.Name)
			}
		}
	}

	// Extract imports
	w.extractImports()

	// Walk all declarations
	for _, decl := range w.file.Decls {
		switch d := decl.(type) {
		case *ast.FuncDecl:
			w.extractFunction(d)
		case *ast.GenDecl:
			w.extractGenDecl(d)
		}
	}
}

func (w *walker) extractImports() {
	for _, imp := range w.file.Imports {
		importPath := strings.Trim(imp.Path.Value, `"`)

		var namespaceImport *string

		if imp.Name != nil {
			if imp.Name.Name == "." {
				// Dot import
				namespaceImport = nil
			} else if imp.Name.Name == "_" {
				ns := "_"
				namespaceImport = &ns
			} else {
				namespaceImport = &imp.Name.Name
			}
		} else {
			parts := strings.Split(importPath, "/")
			last := parts[len(parts)-1]
			namespaceImport = &last
		}

		isExternal := true
		if w.modulePath != "" && strings.HasPrefix(importPath, w.modulePath) {
			isExternal = false
		}

		loc := w.sourceLocation(imp.Pos(), imp.End())

		w.result.Imports = append(w.result.Imports, irtypes.ImportIR{
			Kind:            "import",
			ModulePath:      importPath,
			ResolvedPath:    nil,
			IsExternal:      isExternal,
			Symbols:         []irtypes.ImportedSymbolIR{},
			DefaultImport:   nil,
			NamespaceImport: namespaceImport,
			Location:        loc,
		})
	}
}

func (w *walker) extractFunction(fd *ast.FuncDecl) {
	name := fd.Name.Name
	var qualifiedName *string

	if fd.Recv != nil {
		recvType := receiverTypeName(fd.Recv)
		if recvType != "" {
			qn := recvType + "." + name
			qualifiedName = &qn
		}
	}

	params := extractParams(fd.Type.Params)
	retType := extractReturnType(fd.Type.Results)
	sig := buildSignature(name, fd.Type, fd.Recv)
	isExported := isExportedName(name)

	loc := w.sourceLocation(fd.Pos(), fd.End())

	fnIR := irtypes.FunctionIR{
		Kind:          "function",
		Name:          name,
		QualifiedName: qualifiedName,
		Signature:     sig,
		Parameters:    params,
		ReturnType:    retType,
		IsExported:    isExported,
		IsAsync:       false,
		Location:      loc,
	}

	w.result.Functions = append(w.result.Functions, fnIR)

	if isExported {
		w.result.Exports = append(w.result.Exports, irtypes.ExportIR{
			Kind:       "export",
			Name:       name,
			LocalName:  nil,
			IsDefault:  false,
			FromModule: nil,
			Location:   loc,
		})
	}

	// Walk function body for calls
	prevEnclosing := w.enclosingFunction
	if qualifiedName != nil {
		w.enclosingFunction = *qualifiedName
	} else {
		w.enclosingFunction = name
	}
	if fd.Body != nil {
		ast.Inspect(fd.Body, w.visitCallExpr)
	}
	w.enclosingFunction = prevEnclosing
}

func (w *walker) extractGenDecl(gd *ast.GenDecl) {
	for _, spec := range gd.Specs {
		switch s := spec.(type) {
		case *ast.TypeSpec:
			w.extractTypeSpec(s)
		case *ast.ValueSpec:
			w.extractValueSpec(s)
		}
	}
}

func (w *walker) extractTypeSpec(ts *ast.TypeSpec) {
	name := ts.Name.Name
	isExported := isExportedName(name)
	loc := w.sourceLocation(ts.Pos(), ts.End())

	switch st := ts.Type.(type) {
	case *ast.StructType:
		var superClass *string
		if st.Fields != nil {
			for _, field := range st.Fields.List {
				if len(field.Names) == 0 {
					embeddedName := typeExprToString(field.Type)
					superClass = &embeddedName
					break
				}
			}
		}

		methods := w.structMethods[name]
		if methods == nil {
			methods = []string{}
		}

		w.result.Classes = append(w.result.Classes, irtypes.ClassIR{
			Kind:       "class",
			Name:       name,
			SuperClass: superClass,
			Implements: []string{},
			IsExported: isExported,
			IsAbstract: false,
			Methods:    methods,
			Location:   loc,
		})

	case *ast.InterfaceType:
		var methods []string
		if st.Methods != nil {
			for _, m := range st.Methods.List {
				for _, n := range m.Names {
					methods = append(methods, n.Name)
				}
			}
		}
		if methods == nil {
			methods = []string{}
		}

		w.result.Classes = append(w.result.Classes, irtypes.ClassIR{
			Kind:       "class",
			Name:       name,
			SuperClass: nil,
			Implements: []string{},
			IsExported: isExported,
			IsAbstract: true,
			Methods:    methods,
			Location:   loc,
		})
	}

	if isExported {
		w.result.Exports = append(w.result.Exports, irtypes.ExportIR{
			Kind:       "export",
			Name:       name,
			LocalName:  nil,
			IsDefault:  false,
			FromModule: nil,
			Location:   loc,
		})
	}
}

func (w *walker) extractValueSpec(vs *ast.ValueSpec) {
	for _, name := range vs.Names {
		if isExportedName(name.Name) {
			loc := w.sourceLocation(vs.Pos(), vs.End())
			w.result.Exports = append(w.result.Exports, irtypes.ExportIR{
				Kind:       "export",
				Name:       name.Name,
				LocalName:  nil,
				IsDefault:  false,
				FromModule: nil,
				Location:   loc,
			})
		}
	}
}

func (w *walker) visitCallExpr(n ast.Node) bool {
	ce, ok := n.(*ast.CallExpr)
	if !ok {
		return true
	}

	var callee, receiver, method string
	var receiverPtr, methodPtr *string

	switch fn := ce.Fun.(type) {
	case *ast.SelectorExpr:
		method = fn.Sel.Name
		methodPtr = &method
		receiver = exprToString(fn.X)
		receiverPtr = &receiver
		callee = receiver + "." + method
	case *ast.Ident:
		callee = fn.Name
	default:
		callee = exprToString(ce.Fun)
	}

	argCount := len(ce.Args)
	var argRefs []string
	var stringArgs []string

	for _, arg := range ce.Args {
		switch a := arg.(type) {
		case *ast.Ident:
			argRefs = append(argRefs, a.Name)
		case *ast.BasicLit:
			if a.Kind == token.STRING {
				val := strings.Trim(a.Value, `"`)
				val = strings.Trim(val, "`")
				stringArgs = append(stringArgs, val)
			}
		}
	}

	loc := w.sourceLocation(ce.Pos(), ce.End())

	var enclosing *string
	if w.enclosingFunction != "" {
		enc := w.enclosingFunction
		enclosing = &enc
	}

	callIR := irtypes.CallIR{
		Kind:              "call",
		Callee:            callee,
		Receiver:          receiverPtr,
		Method:            methodPtr,
		ArgumentCount:     argCount,
		ArgumentRefs:      argRefs,
		StringArgs:        stringArgs,
		EnclosingFunction: enclosing,
		Location:          loc,
	}

	// Run plugins
	if w.registry != nil {
		if result := w.registry.AnalyzeCall(ce, callee, receiver, method, stringArgs); result != nil {
			if result.Enrichment != nil {
				callIR.Enrichments = append(callIR.Enrichments, *result.Enrichment)
			}
			if result.EndpointInfo != nil {
				w.enrichEnclosingFunction(result.EndpointInfo, result.Enrichment)
			}
		}
	}

	w.result.Calls = append(w.result.Calls, callIR)
	return true
}

func (w *walker) enrichEnclosingFunction(ep *irtypes.EndpointInfo, enrichment *irtypes.Enrichment) {
	if w.enclosingFunction == "" {
		return
	}
	for i := range w.result.Functions {
		fn := &w.result.Functions[i]
		funcName := fn.Name
		if fn.QualifiedName != nil {
			funcName = *fn.QualifiedName
		}
		if funcName == w.enclosingFunction {
			fn.EndpointInfo = ep
			if enrichment != nil {
				fn.Enrichments = append(fn.Enrichments, *enrichment)
			}
			break
		}
	}
}

func (w *walker) sourceLocation(pos, end token.Pos) irtypes.SourceLocation {
	start := w.fset.Position(pos)
	e := w.fset.Position(end)
	return irtypes.SourceLocation{
		StartLine:   start.Line,
		EndLine:     e.Line,
		StartColumn: start.Column,
		EndColumn:   e.Column,
	}
}

// Helper functions

func isExportedName(name string) bool {
	if name == "" {
		return false
	}
	return unicode.IsUpper(rune(name[0]))
}

func receiverTypeName(recv *ast.FieldList) string {
	if recv == nil || len(recv.List) == 0 {
		return ""
	}
	return typeExprToString(recv.List[0].Type)
}

func typeExprToString(expr ast.Expr) string {
	switch t := expr.(type) {
	case *ast.Ident:
		return t.Name
	case *ast.StarExpr:
		return typeExprToString(t.X)
	case *ast.SelectorExpr:
		return exprToString(t.X) + "." + t.Sel.Name
	case *ast.ArrayType:
		return "[]" + typeExprToString(t.Elt)
	case *ast.MapType:
		return "map[" + typeExprToString(t.Key) + "]" + typeExprToString(t.Value)
	case *ast.InterfaceType:
		return "interface{}"
	case *ast.Ellipsis:
		return "..." + typeExprToString(t.Elt)
	default:
		return fmt.Sprintf("%T", expr)
	}
}

func exprToString(expr ast.Expr) string {
	switch e := expr.(type) {
	case *ast.Ident:
		return e.Name
	case *ast.SelectorExpr:
		return exprToString(e.X) + "." + e.Sel.Name
	case *ast.CallExpr:
		return exprToString(e.Fun) + "()"
	case *ast.StarExpr:
		return "*" + exprToString(e.X)
	default:
		return fmt.Sprintf("%T", expr)
	}
}

func extractParams(fields *ast.FieldList) []irtypes.ParameterIR {
	if fields == nil {
		return []irtypes.ParameterIR{}
	}
	var params []irtypes.ParameterIR
	for _, field := range fields.List {
		typeStr := typeExprToString(field.Type)
		isVariadic := false
		if _, ok := field.Type.(*ast.Ellipsis); ok {
			isVariadic = true
		}

		if len(field.Names) == 0 {
			params = append(params, irtypes.ParameterIR{
				Name:           "",
				TypeAnnotation: &typeStr,
				HasDefault:     false,
				IsRest:         isVariadic,
			})
		} else {
			for _, name := range field.Names {
				params = append(params, irtypes.ParameterIR{
					Name:           name.Name,
					TypeAnnotation: &typeStr,
					HasDefault:     false,
					IsRest:         isVariadic,
				})
			}
		}
	}
	return params
}

func extractReturnType(fields *ast.FieldList) *string {
	if fields == nil || len(fields.List) == 0 {
		return nil
	}
	var types []string
	for _, field := range fields.List {
		t := typeExprToString(field.Type)
		if len(field.Names) > 0 {
			for range field.Names {
				types = append(types, t)
			}
		} else {
			types = append(types, t)
		}
	}
	if len(types) == 1 {
		return &types[0]
	}
	result := "(" + strings.Join(types, ", ") + ")"
	return &result
}

func buildSignature(name string, ft *ast.FuncType, recv *ast.FieldList) string {
	var sb strings.Builder
	sb.WriteString("func ")

	if recv != nil && len(recv.List) > 0 {
		sb.WriteString("(")
		r := recv.List[0]
		if len(r.Names) > 0 {
			sb.WriteString(r.Names[0].Name)
			sb.WriteString(" ")
		}
		sb.WriteString(typeExprToString(r.Type))
		sb.WriteString(") ")
	}

	sb.WriteString(name)
	sb.WriteString("(")

	if ft.Params != nil {
		for i, field := range ft.Params.List {
			if i > 0 {
				sb.WriteString(", ")
			}
			typeStr := typeExprToString(field.Type)
			if len(field.Names) > 0 {
				names := make([]string, len(field.Names))
				for j, n := range field.Names {
					names[j] = n.Name
				}
				sb.WriteString(strings.Join(names, ", "))
				sb.WriteString(" ")
			}
			sb.WriteString(typeStr)
		}
	}

	sb.WriteString(")")

	if ft.Results != nil && len(ft.Results.List) > 0 {
		sb.WriteString(" ")
		if len(ft.Results.List) > 1 || len(ft.Results.List[0].Names) > 0 {
			sb.WriteString("(")
		}
		for i, field := range ft.Results.List {
			if i > 0 {
				sb.WriteString(", ")
			}
			if len(field.Names) > 0 {
				names := make([]string, len(field.Names))
				for j, n := range field.Names {
					names[j] = n.Name
				}
				sb.WriteString(strings.Join(names, ", "))
				sb.WriteString(" ")
			}
			sb.WriteString(typeExprToString(field.Type))
		}
		if len(ft.Results.List) > 1 || len(ft.Results.List[0].Names) > 0 {
			sb.WriteString(")")
		}
	}

	return sb.String()
}
